### PR TITLE
Implement faithful representations of nilpotent Lie algebras

### PR DESCRIPTION
**Pull request by tscrim**
We implement two methods to construct faithful representations of nilpotent Lie algebras. This is a first step towards implementing Ado's theorem.

### :memo: Checklist

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/algebras/lie_algebras/representation.py [hunk 1]
```diff
@@ -433,3 +474,313 @@ def _acted_upon_(self, scalar, self_on_left=False):
...
+    The resulting quotient `U(L) / I` remains faithful and is a minimal
+    faithful representation of `L` in the sense that it has no faithful
+    submodules or quotients. (Note: this is not necessarily the smallest
+    dimensional faithful representation of `L`.)
+
+    We consider an example of the rank 2 Heisenberg Lie algebra,
+    but with a non-standard basis given by `a = p_1 + z`, `b = q_1`,
+    and `c = q_1 + z`::
+
+        sage: L.<a,b,c> = LieAlgebra(QQ, {('a','b'): {'b':-1, 'c':1}, ('a','c'): {'b':-1, 'c':1}})
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/representation.py]:
```suggestion
        sage: L.<a,b,c> = LieAlgebra(QQ, {('a','b'): {'b':-1, 'c':1},
        ....:                             ('a','c'): {'b':-1, 'c':1}})
```

File: src/sage/algebras/lie_algebras/representation.py [hunk 2]
```diff
@@ -433,3 +474,313 @@ def _acted_upon_(self, scalar, self_on_left=False):
...
+        ]
+
+    An example with ``minimal=True`` for `H_2 \oplus A_1`, where `A_1` is
+    a `1`-dimensional Abelian Lie algebra::
+
+        sage: L.<a,b,c,d> = LieAlgebra(QQ, {('a','b'): {'b':-1, 'c':1}, ('a','c'): {'b':-1, 'c':1}})
+        sage: F = L.faithful_representation(); F
+        Faithful 11 dimensional representation of Lie algebra on 4 generators (a, b, c, d) over Rational Field
+        sage: MF = L.faithful_representation(algorithm="minimal"); MF
+        Minimal faithful representation of Lie algebra on 4 generators (a, b, c, d) over Rational Field
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/representation.py]:
```suggestion
        sage: L.<a,b,c,d> = LieAlgebra(QQ, {('a','b'): {'b':-1, 'c':1}, 
        ....:                               ('a','c'): {'b':-1, 'c':1}})
        sage: F = L.faithful_representation(); F
        Faithful 11 dimensional representation of
         Lie algebra on 4 generators (a, b, c, d) over Rational Field
        sage: MF = L.faithful_representation(algorithm="minimal"); MF
        Minimal faithful representation of
         Lie algebra on 4 generators (a, b, c, d) over Rational Field
```

File: src/sage/algebras/lie_algebras/representation.py [hunk 3]
```diff
@@ -433,3 +474,313 @@ def _acted_upon_(self, scalar, self_on_left=False):
...
+
+        EXAMPLES::
+
+            sage: H2 = lie_algebras.Heisenberg(QQ, 2)
+            sage: F = H2.faithful_representation()
+            sage: TestSuite(F).run(elements=list(F.basis()))
+            sage: MF = H2.faithful_representation(algorithm="minimal")
+            sage: TestSuite(MF).run(elements=list(MF.basis()))
+
+            sage: L.<a,b,c> = LieAlgebra(QQ, {('a','b'): {'b':-1, 'c':1}, ('a','c'): {'b':-1, 'c':1}})
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/representation.py]:
```suggestion
            sage: L.<a,b,c> = LieAlgebra(QQ, {('a','b'): {'b':-1, 'c':1},                             
            ....:                             ('a','c'): {'b':-1, 'c':1}})
```

File: src/sage/algebras/lie_algebras/representation.py [hunk 4]
```diff
@@ -433,3 +474,313 @@ def _acted_upon_(self, scalar, self_on_left=False):
...
+            ZB = [L(b) for b in Z.basis()]
+            prev = LCS[-1]
+            for D in reversed(LCS[:-1]):
+                cur = []
+                for ind in range(len(ZB)-1, -1, -1):
+                    z = ZB[ind]
+                    if z in D:
+                        ZB.pop(ind)
+                        cur.append(z)
+                k = self._step-len(basis_by_deg)
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/representation.py]:
```suggestion
                k = self._step - len(basis_by_deg)
```

File: src/sage/algebras/lie_algebras/representation.py [hunk 5]
```diff
@@ -433,3 +474,313 @@ def _acted_upon_(self, scalar, self_on_left=False):
...
+        basis_by_deg = {}
+        self._step = len(LCS) - 1
+        self._minimal = minimal
+        if self._minimal:
+            Z = L.center()
+            ZB = [L(b) for b in Z.basis()]
+            prev = LCS[-1]
+            for D in reversed(LCS[:-1]):
+                cur = []
+                for ind in range(len(ZB)-1, -1, -1):
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/representation.py]:
```suggestion
                for ind in range(len(ZB) - 1, -1, -1):
```

File: src/sage/algebras/lie_algebras/representation.py [hunk 6]
```diff
@@ -433,3 +474,313 @@ def _acted_upon_(self, scalar, self_on_left=False):
...
+            return "\\min " + ret
+        return ret
+
+    def _project(self, elt):
+        r"""
+        The projection to ``self`` from the PBW basis.
+
+        EXAMPLES::
+
+            sage: L.<a,b,c> = LieAlgebra(QQ, {('a','b'): {'b':-1, 'c':1}, ('a','c'): {'b':-1, 'c':1}})
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/representation.py]:
```suggestion
            sage: L.<a,b,c> = LieAlgebra(QQ, {('a','b'): {'b':-1, 'c':1},                             
            ....:                             ('a','c'): {'b':-1, 'c':1}})
```

File: src/sage/algebras/lie_algebras/representation.py [hunk 7]
```diff
@@ -433,3 +474,313 @@ def _acted_upon_(self, scalar, self_on_left=False):
...
+                    ret[self._indices(vec)] = c
+        return self.element_class(self, ret)
+
+    def _to_pbw(self, elt):
+        """
+        Return the PBW element corresponding to ``elt``.
+
+        EXAMPLES::
+
+            sage: L.<a,b,c> = LieAlgebra(QQ, {('a','b'): {'b':-1, 'c':1}, ('a','c'): {'b':-1, 'c':1}})
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/representation.py]:
```suggestion
            sage: L.<a,b,c> = LieAlgebra(QQ, {('a','b'): {'b':-1, 'c':1},                             
            ....:                             ('a','c'): {'b':-1, 'c':1}})
```

File: src/sage/categories/finite_dimensional_lie_algebras_with_basis.py [hunk 8]
```diff
@@ -1927,6 +1927,88 @@ def to_prod(vec, index):
...
+
+            EXAMPLES::
+
+                sage: H2 = lie_algebras.Heisenberg(QQ, 2)
+                sage: H2.is_nilpotent()
+                True
+                sage: F = H2.faithful_representation(); F
+                Faithful 16 dimensional representation of Heisenberg algebra of rank 2 over Rational Field
+                sage: M = H2.faithful_representation(algorithm="minimal"); M
+                Minimal faithful representation of Heisenberg algebra of rank 2 over Rational Field
```
Comment (mkoeppe) [inline src/sage/categories/finite_dimensional_lie_algebras_with_basis.py]:
```suggestion
                sage: F = H2.faithful_representation(); F
                Faithful 16 dimensional representation of 
                 Heisenberg algebra of rank 2 over Rational Field
                sage: M = H2.faithful_representation(algorithm="minimal"); M
                Minimal faithful representation of 
                 Heisenberg algebra of rank 2 over Rational Field
```

**Comment by tscrim** [comment]
Last commit should address all of the comments.

File: src/sage/algebras/lie_algebras/representation.py [hunk 9]
```diff
@@ -94,6 +97,44 @@ def _test_representation(self, **options):
             for v in S:
                 tester.assertEqual(x.bracket(y) * v, x * (y * v) - y * (x * v))
 
+    def representation_matrix(self, elt):
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/representation.py:100]: The method of this name in `sage.modules.with_basis.representation` has additional keyword arguments, including `side`. Should this one too?

**Review by mkoeppe — APPROVED** [review]
Otherwise LGTM.

File: src/sage/algebras/lie_algebras/representation.py (hunk 9 above)
Comment (tscrim) [reply src/sage/algebras/lie_algebras/representation.py:100]: No, the action is only implemented from the left. So it doesn't make sense to have it.

Comment (mkoeppe) [reply src/sage/algebras/lie_algebras/representation.py:100]: OK then

**Comment by tscrim** [comment]
Thank you.

**Pull request merged**